File: dotnet/test/Azure.Iot.Operations.Protocol.MetlTests/TestCaseActionReceiveTelemetry.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Azure.Iot.Operations.Protocol.MetlTests
{
    public class TestCaseActionReceiveTelemetry : TestCaseAction
    {
        public static string? DefaultTopic;
        public static string? DefaultPayload;
        public static string? DefaultContentType;
        public static int? DefaultFormatIndicator;
        public static int? DefaultQos;
        public static TestCaseDuration? DefaultMessageExpiry;
        public static int? DefaultSourceIndex;

        public string? Topic { get; set; } = DefaultTopic;

        public string? Payload { get; set; } = DefaultPayload;

        public bool BypassSerialization { get; set; }

        public string? ContentType { get; set; } = DefaultContentType;

        public int? FormatIndicator { get; set; } = DefaultFormatIndicator;

        public Dictionary<string, string> Metadata { get; set; } = new();

        public int? Qos { get; set; } = DefaultQos;

        public TestCaseDuration? MessageExpiry { get; set; } = DefaultMessageExpiry;

        public int? SourceIndex { get; set; } = DefaultSourceIndex;

        public int? PacketIndex { get; set; }
    }
}

File: go/test/protocol/test_case_action_receive_telemetry.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package protocol

import (
	"gopkg.in/yaml.v3"
)

type testCaseActionReceiveTelemetry struct {
	Topic           *string           `yaml:"topic"`
	Payload         *string           `yaml:"payload"`
	ContentType     *string           `yaml:"content-type"`
	FormatIndicator *int              `yaml:"format-indicator"`
	Metadata        map[string]string `yaml:"metadata"`
	Qos             *int              `yaml:"qos"`
	MessageExpiry   *TestCaseDuration `yaml:"message-expiry"`
	SourceIndex     *int              `yaml:"source-index"`
	PacketIndex     *int              `yaml:"packet-index"`
}

type TestCaseActionReceiveTelemetry struct {
	testCaseActionReceiveTelemetry
}

func (receiveTelemetry *TestCaseActionReceiveTelemetry) UnmarshalYAML(
	node *yaml.Node,
) error {
	*receiveTelemetry = TestCaseActionReceiveTelemetry{}

	receiveTelemetry.Topic = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetTopic()
	receiveTelemetry.Payload = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetPayload()
	receiveTelemetry.ContentType = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetContentType()
	receiveTelemetry.FormatIndicator = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetFormatIndicator()
	receiveTelemetry.Qos = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetQos()
	receiveTelemetry.SourceIndex = TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetSourceIndex()

	defaultMessageExpiry := TestCaseDefaultInfo.Actions.ReceiveTelemetry.GetMessageExpiry()
	receiveTelemetry.MessageExpiry = &defaultMessageExpiry

	return node.Decode(&receiveTelemetry.testCaseActionReceiveTelemetry)
}
